### PR TITLE
rlp: Fix buffer indexing

### PR DESCRIPTION
**Pull request by vorot93**
Users can provide their own buffer now - we must assume that buffers start with arbitrary length.

### CODE REVIEW

**Review by ordian — APPROVED** [review]
Could you add a test for `clear` as well?

File: rlp/src/stream.rs [hunk 1]
```diff
@@ -359,11 +369,11 @@ impl<'a> BasicEncoder<'a> {
 		// 1 byte was already reserved for payload earlier
 		match len {
 			0..=55 => {
-				self.buffer[pos - 1] = 0xc0u8 + len as u8;
+				self.buffer[self.start_pos + pos - 1] = 0xc0u8 + len as u8;
```
Comment (dvdplm) [inline rlp/src/stream.rs:372]: If `pos` is ever zero it looks like nasty things could happen (was true before too); can you double-check it cannot be called with `pos` == 0?

File: rlp/src/stream.rs [hunk 2]
```diff
@@ -403,9 +413,9 @@ impl<'a> BasicEncoder<'a> {
 			// (prefix + length of length), followed by the length, followd by the string
 			len => {
 				self.buffer.put_u8(0);
-				let position = self.buffer.len();
+				let position = self.total_written();
 				let inserted_bytes = self.insert_size(len, position);
-				self.buffer[position - 1] = 0xb7 + inserted_bytes;
+				self.buffer[self.start_pos + position - 1] = 0xb7 + inserted_bytes;
```
Comment (dvdplm) [inline rlp/src/stream.rs:418]: Same concern here, if both `start_pos` and `position` can be `0` (and `self.total_written()` starts at `0` right?) then we'd get an out of bounds panic right?

File: rlp/src/stream.rs (hunk 1 above)
Comment (vorot93) [reply rlp/src/stream.rs:372]: I cannot be sure without auditing all possible encoding paths in this codebase, but this is orthogonal to this PR as it _increases_ the count.

I suggest we merge as is, and add this concern to a round of fuzzing which this library so badly needs anyway.

File: rlp/src/stream.rs (hunk 2 above)
Comment (vorot93) [reply rlp/src/stream.rs:418]: see above

File: rlp/src/stream.rs (hunk 1 above)
Comment (niklasad1) [reply rlp/src/stream.rs:372]: Yepp, agree unrelated to this PR but we should open a ticket for it and investigate it further.

**Review by niklasad1 — APPROVED** [review]

**Pull request merged**